File: lua/test/custom_entity_test.lua
-- Custom entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("codatplatform_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CustomEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Custom(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = custom_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "custom." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set CODATPLATFORM_TEST_CUSTOM_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local custom_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.custom")))
    local custom_ref01_data = nil
    if #custom_ref01_data_raw > 0 then
      custom_ref01_data = helpers.to_map(custom_ref01_data_raw[1][2])
    end

    -- UPDATE
    local custom_ref01_ent = client:Custom(nil)
    local custom_ref01_data_up0_up = {
      ["platform_key"] = setup.idmap["platform_key"],
    }

    local custom_ref01_markdef_up0_name = "dataSource"
    local custom_ref01_markdef_up0_value = "Mark01-custom_ref01_" .. tostring(setup.now)
    custom_ref01_data_up0_up[custom_ref01_markdef_up0_name] = custom_ref01_markdef_up0_value

    local custom_ref01_resdata_up0_result, err = custom_ref01_ent:update(custom_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local custom_ref01_resdata_up0 = helpers.to_map(type(custom_ref01_resdata_up0_result) == 'table' and custom_ref01_resdata_up0_result.data_get and custom_ref01_resdata_up0_result:data_get() or custom_ref01_resdata_up0_result)
    assert.is_not_nil(custom_ref01_resdata_up0)
    assert.are.equal(custom_ref01_resdata_up0[custom_ref01_markdef_up0_name], custom_ref01_markdef_up0_value)

    -- LOAD
    local custom_ref01_match_dt0 = {}
    local custom_ref01_data_dt0_loaded, err = custom_ref01_ent:load(custom_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(custom_ref01_data_dt0_loaded)

  end)
end)

function custom_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/custom/CustomTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read custom test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "custom01", "custom02", "custom03", "integration01", "integration02", "integration03", "company01", "company02", "company03", "connection01", "connection02", "connection03", "platform_key01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("CODATPLATFORM_TEST_CUSTOM_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["CODATPLATFORM_TEST_CUSTOM_ENTID"] = idmap,
    ["CODATPLATFORM_TEST_LIVE"] = "FALSE",
    ["CODATPLATFORM_TEST_EXPLAIN"] = "FALSE",
    ["CODATPLATFORM_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["CODATPLATFORM_TEST_CUSTOM_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["platform_key"] == nil then
    idmap_resolved["platform_key"] = idmap_resolved["platform_key01"]
  end

  if env["CODATPLATFORM_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["CODATPLATFORM_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["CODATPLATFORM_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["CODATPLATFORM_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/custom_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/codatplatform-sdk/go"
	"github.com/voxgig-sdk/codatplatform-sdk/go/core"

	vs "github.com/voxgig-sdk/codatplatform-sdk/go/utility/struct"
)

func TestCustomEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Custom(nil)
		if ent == nil {
			t.Fatal("expected non-nil CustomEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := customBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "custom." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CODATPLATFORM_TEST_CUSTOM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		customRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.custom", setup.data)))
		var customRef01Data map[string]any
		if len(customRef01DataRaw) > 0 {
			customRef01Data = core.ToMapAny(customRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = customRef01Data

		// UPDATE
		customRef01Ent := client.Custom(nil)
		customRef01DataUp0Up := map[string]any{
			"id": customRef01Data["id"],
			"platform_key": setup.idmap["platform_key"],
		}

		customRef01MarkdefUp0Name := "dataSource"
		customRef01MarkdefUp0Value := fmt.Sprintf("Mark01-custom_ref01_%d", setup.now)
		customRef01DataUp0Up[customRef01MarkdefUp0Name] = customRef01MarkdefUp0Value

		customRef01ResdataUp0Result, err := customRef01Ent.Update(customRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		customRef01ResdataUp0 := core.ToMapAny(entityData(customRef01ResdataUp0Result))
		if customRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if customRef01ResdataUp0["id"] != customRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if customRef01ResdataUp0[customRef01MarkdefUp0Name] != customRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", customRef01MarkdefUp0Name, customRef01ResdataUp0[customRef01MarkdefUp0Name])
		}

		// LOAD
		customRef01MatchDt0 := map[string]any{
			"id": customRef01Data["id"],
		}
		customRef01DataDt0Loaded, err := customRef01Ent.Load(customRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		customRef01DataDt0LoadResult := core.ToMapAny(entityData(customRef01DataDt0Loaded))
		if customRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if customRef01DataDt0LoadResult["id"] != customRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func customBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "custom", "CustomTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read custom test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse custom test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"custom01", "custom02", "custom03", "integration01", "integration02", "integration03", "company01", "company02", "company03", "connection01", "connection02", "connection03", "platform_key01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CODATPLATFORM_TEST_CUSTOM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CODATPLATFORM_TEST_CUSTOM_ENTID": idmap,
		"CODATPLATFORM_TEST_LIVE":      "FALSE",
		"CODATPLATFORM_TEST_EXPLAIN":   "FALSE",
		"CODATPLATFORM_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CODATPLATFORM_TEST_CUSTOM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add platform_key alias for update test.
	if idmapResolved["platform_key"] == nil {
		idmapResolved["platform_key"] = idmapResolved["platform_key01"]
	}

	if env["CODATPLATFORM_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CODATPLATFORM_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCodatplatformSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CODATPLATFORM_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CODATPLATFORM_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
